lib/codecs/src/encoding/format/csv.rs

```diff
@@ -0,0 +1,184 @@
+use crate::encoding::BuildError;
+use bytes::{BufMut, BytesMut};
+use lookup::lookup_v2::ConfigOwnedValuePath;
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, Value},
+    schema,
+};
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerConfig {
+    /// The CSV Serializer Options.
+    pub csv: CsvSerializerOptions,
+}
+
+impl CsvSerializerConfig {
+    /// Creates a new `CsvSerializerConfig`.
+    pub const fn new(csv: CsvSerializerOptions) -> Self {
+        Self { csv }
+    }
+
+    /// Build the `CsvSerializer` from this configuration.
+    pub fn build(&self) -> Result<CsvSerializer, BuildError> {
+        if self.csv.fields.is_empty() {
+            Err("At least one CSV field must be specified".into())
+        } else {
+            Ok(CsvSerializer::new(self.csv.fields.clone()))
+        }
+    }
+
+    /// The data type of events that are accepted by `CsvSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // CSV, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerOptions {
+    /// The CSV Serializer fields.
+    ///
+    /// The order of the fields determines the order of the columns in the CSV output.
+    /// Make sure to use the same order when parsing the CSV output.
+    ///
+    /// If a field is not present in the event, the output will be `NaN`.
+    ///
+    /// Some value types: Array, Regex, Object are not supported by the CSV format.
+    /// If a field is of an unsupported type, the output will be `NaN`.
+    pub fields: Vec<ConfigOwnedValuePath>,
```

Comment:
Just making sure the choice of `ValuePath` here was intentional, instead of `TargetPath`. This will prevent event metadata from being serialized. I'm not seeing a great reason to restrict it.

Reply:
Note: If this is switched, an equivalent `ConfigOwnedTargetPath` doesn't exist yet, but I can quickly make one.

Reply:
It seems OK to change it to `TargetPath`. I chose `ConfigOwnedValuePath` since it is also used in `Transformer`.

Reply:
I think that's correct, but @fuchsnj can you confirm?

Reply:
`ConfigOwnedTargetPath` shouldn't have `Default` either. `CsvSerializerOptions` and `CsvSerializerConfig` can have `Default` removed since they are unused / also not needed.